base/setup/lib/utils/filesup.c

```diff
@@ -71,79 +66,80 @@ SetupCreateSingleDirectory(
                           NULL,
                           0);
     if (NT_SUCCESS(Status))
-    {
         NtClose(DirectoryHandle);
-    }
-
-    RtlFreeUnicodeString(&PathName);
 
     return Status;
 }
 
+/**
+ * @brief
+ * Create a new directory, specified by the given path.
+ * Any intermediate non-existing directory is created as well.
+ *
+ * @param[in]   PathName
+ * The path of the directory to be created.
+ *
+ * @return  An NTSTATUS code indicating success or failure.
+ **/
 NTSTATUS
 SetupCreateDirectory(
```

Comment:
If you create a UNICODE_STRING version of this, it could be implemented even simpler:
```
SetupCreateDirectory(U_S* str)
{
	if (RtlFindCharInUnicodeString (str, RTL_FIND_CHAR_IN_UNICODE_STRING_START_AT_END, OBJ_NAME_PATH_SEPARATOR, &PathDivider, ...))
	{
		UNICODE_STRING parent = *str - PathDivider;
		if (Exists(parent))
			return;

		SetupCreateDirectory(&parent);
	}
	else
	{
		// done
	}
}
```

Reply:
Yep; except I avoid using recursion here, so I could still use the RtlFindChar... in a loop.